runtime/src/main/java/org/corfudb/runtime/view/BatchSequencerView.java

```diff
@@ -0,0 +1,252 @@
+package org.corfudb.runtime.view;
+
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Table;
+import com.google.common.hash.HashCode;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import lombok.Data;
+import org.corfudb.protocols.wireprotocol.BatchTokenRequest;
+import org.corfudb.protocols.wireprotocol.BatchTokenResponse;
+import org.corfudb.protocols.wireprotocol.BatchTokenResponse.BackpointerToken;
+import org.corfudb.protocols.wireprotocol.Token;
+import org.corfudb.protocols.wireprotocol.TokenResponse;
+import org.corfudb.protocols.wireprotocol.TokenType;
+import org.corfudb.protocols.wireprotocol.TxResolutionInfo;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+
+public class BatchSequencerView extends SequencerView {
+
+    @Data
+    private static class BatchSequencerOperation {
+
+        final boolean next;
+
+        final TxResolutionInfo info;
+
+        final UUID[] streams;
+
+        final CompletableFuture<TokenResponse> completion = new CompletableFuture<>();
+
+        TokenResponse response = null;
+
+        void completeResponse() {
+            if (response == null) {
+                throw new UnrecoverableCorfuError("Attempted to complete an empty token");
+            }
+            completion.complete(response);
+        }
+    }
+
+    Semaphore s = new Semaphore(0);
+
+    final Deque<BatchSequencerOperation> queue = new ConcurrentLinkedDeque<>();
+
+    final ExecutorService batcher = Executors.newFixedThreadPool(1, new ThreadFactoryBuilder()                                                  .setNameFormat("sequencer-batcher-%d").build());
+
+    volatile long lastGlobal = Address.NEVER_READ;
+    volatile long lastEpoch = -1;
+
+    public BatchSequencerView(CorfuRuntime runtime) {
+        super(runtime);
+        batcher.submit(this::sequencerBatcher);
+    }
+
+    private void sequencerBatcher() {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 52 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)